app/src/main/java/mozilla/lockbox/autofill/FillResponseBuilder.kt

```diff
@@ -0,0 +1,103 @@
+package mozilla.lockbox.autofill
+
+import android.annotation.TargetApi
+import android.content.Context
+import android.os.Build
+import android.service.autofill.Dataset
+import android.service.autofill.FillResponse
+import android.view.autofill.AutofillValue
+import android.widget.RemoteViews
+import mozilla.appservices.logins.ServerPassword
+import mozilla.lockbox.R
+import mozilla.lockbox.support.ParsedStructure
+import mozilla.lockbox.view.AuthActivity
+import java.net.URL
+
+@TargetApi(Build.VERSION_CODES.O)
+class FillResponseBuilder(
+    private val parsedStructure: ParsedStructure,
+    private val hostname: String
+) {
+    fun buildAuthenticationFillResponse(context: Context): FillResponse {
+        val responseBuilder = FillResponse.Builder()
+
+        val authPresentation = RemoteViews(context.packageName, R.layout.autofill_item).apply {
+            val callToAction = context.resources.getString(R.string.autofill_authenticate_cta)
+            setTextViewText(R.id.presentationText, callToAction)
+        }
+
+        val sender = AuthActivity.getAuthIntentSender(context, this)
+        val autofillIds = arrayOf(parsedStructure.usernameId, parsedStructure.passwordId)
+
+        responseBuilder.setAuthentication(autofillIds, sender, authPresentation)
+
+        return responseBuilder.build()
+    }
+
+    fun buildFallbackFillResponse(context: Context): FillResponse? {
+        // See https://github.com/mozilla-lockbox/lockbox-android/issues/421
+        val builder = FillResponse.Builder()
+        addSearchFooter(context, builder)
+        return builder.build()
+    }
+
+    private fun addSearchFooter(context: Context, builder: FillResponse.Builder) {
+        val searchPresentation = RemoteViews(context.packageName, R.layout.autofill_item).apply {
+            val callToAction = context.resources.getString(R.string.autofill_search_cta)
+            setTextViewText(R.id.presentationText, callToAction)
+        }
+
+        // See https://github.com/mozilla-lockbox/lockbox-android/issues/421
+        val sender = AuthActivity.getAuthIntentSender(context, this)
+        val autofillIds = arrayOf(parsedStructure.usernameId, parsedStructure.passwordId)
+
+        builder.setAuthentication(autofillIds, sender, searchPresentation)
+    }
+
+    fun buildFilteredFillResponse(context: Context, passwords: List<ServerPassword>): FillResponse? {
+        val possibleValues = passwords.filter {
+            val credential = URL(it.hostname)
+            (credential.host.endsWith(hostname, true) || hostname.endsWith(credential.host, true))
```

Comment:
I think some of this logic will change / need to incorporate @linuxwolf 's work at #425

Reply:
Yis.